src/core/leader_schedule.zig

```diff
@@ -8,206 +8,241 @@ const Epoch = sig.core.Epoch;
 const Pubkey = sig.core.Pubkey;
 const Slot = sig.core.Slot;
 const WeightedRandomSampler = sig.rand.WeightedRandomSampler;
-const RpcClient = sig.rpc.Client;
+const EpochSchedule = sig.core.EpochSchedule;
 
 pub const NUM_CONSECUTIVE_LEADER_SLOTS: u64 = 4;
+pub const MAX_CACHED_LEADER_SCHEDULES: usize = 10;
 
 pub const SlotLeaderProvider = sig.utils.closure.PointerClosure(Slot, ?Pubkey);
 
-/// Only works for a single epoch. This is a basic limited approach that should
-/// only be used as a placeholder until a better approach is fleshed out.
-pub const SingleEpochLeaderSchedule = struct {
-    allocator: Allocator,
-    slot_leaders: []const sig.core.Pubkey,
-    start_slot: Slot,
+/// LeaderScheduleCache is a cache of leader schedules for each epoch.
+/// Leader schedules are expensive to compute, so this cache is used to avoid
+/// recomputing leader schedules for the same epoch.
+/// LeaderScheduleCache also keeps a copy of the epoch_schedule so that it can
+/// compute epoch and slot index from a slot.
+pub const LeaderScheduleCache = struct {
+    allocator: std.mem.Allocator,
+    epoch_schedule: EpochSchedule,
+    rwlock: std.Thread.RwLock,
+    leader_schedules: std.AutoArrayHashMapUnmanaged(Epoch, LeaderSchedule),
 
     const Self = @This();
 
-    pub fn deinit(self: Self) void {
-        self.allocator.free(self.slot_leaders);
+    pub fn init(allocator: Allocator, epoch_schedule: EpochSchedule) Self {
+        return .{
+            .allocator = allocator,
+            .epoch_schedule = epoch_schedule,
+            .rwlock = .{},
+            .leader_schedules = .{},
+        };
     }
 
-    pub fn getLeader(self: *const Self, slot: Slot) ?sig.core.Pubkey {
-        const index: usize = @intCast(slot - self.start_slot);
-        return if (index >= self.slot_leaders.len) null else self.slot_leaders[index];
+    pub fn slotLeaderProvider(self: *Self) SlotLeaderProvider {
+        return SlotLeaderProvider.init(self, LeaderScheduleCache.slotLeader);
     }
 
-    pub fn provider(self: *Self) SlotLeaderProvider {
-        return SlotLeaderProvider.init(self, Self.getLeader);
+    pub fn put(self: *Self, epoch: Epoch, leader_schedule: LeaderSchedule) !void {
+        self.rwlock.lock();
+        defer self.rwlock.unlock();
```

Comment:
do we want to respect `MAX_CACHED_LEADER_SCHEDULES` here?

Reply:
d055f2

Reply:
c916da